content/6_Advanced/Segtree_Beats.mdx

```diff
@@ -1,21 +1,531 @@
+**Modulo either does not affect a number, or decreases it by at least half of
+what it was**. If the number in question is $x$, and the modulo was by $m$, then
+this can be proved using casework:
+
+- If $m > x$, then $x$ is uneffected by $m$
+- If $m \le x / 2$, then after the modulo operation $x$ must be strictly less than
+  $m$
+- If $x / 2 < m \leq x$, then $x \mod m = x - m$. This then reduces to
+  the second case.
+
+Let us ignore operations of type 3 for the time being. Because of this property
+of modulo, an element with value $a$ will get decreased at most
+$\lceil \log a \rceil$ times (although a greater number of updates may not
+affect the element). Taking this into account, we can slightly modify the modulo
+update function to encorporate these optimizations.
+
+```
+function update(upd_left, upd_right, upd_value, tree_node, tree_left, tree_right)
+	let cur_max = maximum value in current node
+	let cur_min = minimum value in current node
+
+	if upd_right < tree_left or tree_right < query_left or cur_max < upd_value
+		return
+	if query_left ≤ tree_left and tree_right ≤ query_right and cur_max = cur_min
+		apply update
+		return
+	push lazy updates down
+
+	let tree_mid = (tree_left + tree_right) / 2
+	let left_child = 2 * tree_node
+	let right_child = 2 * tree_node + 1
+	update(upd_left, upd_right, upd_value, left_child, tree_left, tree_mid)
+	update(upd_left, upd_right, upd_value, right_child, tree_mid + 1, tree_right)
+	merge values from children
+```
+
+We can store $\texttt{cur\_max}$ and $\texttt{cur\_min}$ to compute them
+quickly (similar to a max/min segment tree). For a given node, if
+$\texttt{cur\_max} = \texttt{cur\_min}$, that means all elements in the range
+are the same, which means the sum can be easily computed and updated. Note that
+it is still necessary to maintain a lazy tag to propagate updates. Clearly, this
+solution runs in $\mathcal{O}((n + q)\log a)$ where $a$ is the maximum initial
+value of an element.
+
+Consider adding in operations of type 3. Although the implementation is
+relatively straightforward, the proof of complexity from the previous section
+falls apart because elements can be increased back to their maximum value.
+However, not all hope is lost.
+
+Define the entropy of the array to be $\sum_{k = 1}^n \lceil \log a_k \rceil$,
+or equivalently, the maximum number of modulo operations to decrease the array
+to its base state of all 0s. If there are no point updates, then the time
+complexity is linear to the entropy, which has a maximum value of $n \log a$.
+Each point update increases the entropy by a fixed amount
+$\lceil \log a \rceil$. If there are $q_p$ point updates, then the total entropy
+over all updates is bounded by $n \log a + q_p \log a$. If we factor out these
+point update operations, each modulo update is still bounded by the total
+entropy. This means that even with point updates, our solution still runs in
+$\mathcal{O}(n \log a + q_p \log a) = \mathcal{O}((n + q)\log a)$.
```

Comment:
I'm not convinced. A point modulo operation will reduce the entropy by only 1, but it will $\log n$ time, so shouldn't this bound be `\mathcal{O}((n + q)\log a\log n)$`? I'm not sure how to actually construct a test case that results in this complexity though.